clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VeniceAvroRecordReader.java

```diff
@@ -82,36 +88,78 @@ public VeniceAvroRecordReader(
-        props.getString(KEY_FIELD_PROP),
-        props.getString(VALUE_FIELD_PROP),
-        ETLValueSchemaTransformation
-            .valueOf(props.getString(ETL_VALUE_SCHEMA_TRANSFORMATION, ETLValueSchemaTransformation.NONE.name())));
-  }
-
   /**
    * This constructor is used in the Mapper.
-   * @param topicName Topic which is to be published to
-   * @param fileSchema Schema of the source files
-   * @param keyFieldStr Field name of the key field
-   * @param valueFieldStr Field name of the value field
-   * @param etlValueSchemaTransformation The type of transformation that was applied to this schema during ETL. When source data set is not an ETL job, use NONE.
    */
-  public VeniceAvroRecordReader(
-      String topicName,
-      Schema fileSchema,
-      String keyFieldStr,
-      String valueFieldStr,
-      ETLValueSchemaTransformation etlValueSchemaTransformation) {
-    super(topicName);
-    this.fileSchema = fileSchema;
-    this.etlValueSchemaTransformation = etlValueSchemaTransformation;
-    setupSchema(keyFieldStr, valueFieldStr);
+
+  public VeniceAvroRecordReader(VeniceProperties props) {
+    super(props.getString(TOPIC_PROP));
+    this.fileSchema = AvroSchemaParseUtils.parseSchemaFromJSON(
+        props.getString(SCHEMA_STRING_PROP),
+        props.getBoolean(EXTENDED_SCHEMA_VALIDITY_CHECK_ENABLED, DEFAULT_EXTENDED_SCHEMA_VALIDITY_CHECK_ENABLED));
+
+    this.etlValueSchemaTransformation = ETLValueSchemaTransformation
+        .valueOf(props.getString(ETL_VALUE_SCHEMA_TRANSFORMATION, ETLValueSchemaTransformation.NONE.name()));
+    this.generatePartialUpdateRecordFromInput = props.getBoolean(PARTIAL_UPDATE_WITH_NEW_INPUT_FORMAT, false);
+    if (generatePartialUpdateRecordFromInput) {
+      this.updateSchema =
+          AvroSchemaParseUtils.parseSchemaFromJSONLooseValidation(props.getString(UPDATE_SCHEMA_STRING_PROP));
+    }
+    setupSchema(props.getString(KEY_FIELD_PROP), props.getString(VALUE_FIELD_PROP));
+  }
+
+  @Override
+  protected Object getAvroKey(AvroWrapper<IndexedRecord> record, NullWritable nullValue) {
+    Object keyDatum = record.datum().get(keyFieldPos);
+
+    if (keyDatum == null) {
+      // Invalid data
+      // Theoretically it should not happen since all the avro records are sharing the same schema in the same file
+      throw new VeniceException("Encountered record with null key");
+    }
+
+    return keyDatum;
+  }
+
+  @Override
+  protected Object getAvroValue(AvroWrapper<IndexedRecord> record, NullWritable nullValue) {
+    Object valueObject = record.datum().get(valueFieldPos);
+    if (!generatePartialUpdateRecordFromInput) {
+      return valueObject;
+    }
+    if (!(valueObject instanceof IndexedRecord)) {
+      throw new VeniceException("Retrieved record is not a Avro indexed record");
+    }
+    UpdateBuilder updateBuilder = new UpdateBuilderImpl(updateSchema);
```

Comment:
We can evaluate reusing `UpdateBuilder` in the future since every instance creation will do the repetitive work, such as schema validation, retrieving fast serde class (which is proved to be slow because of schema comparison).

Reply:
Sure, we can offer that, this should not be too difficult to achieve in my opinion.